bigchaindb/upsert_validator/validator_election.py

```diff
@@ -218,9 +224,58 @@ def get_validator_update(cls, bigchain, new_height, txns):
                 validator_updates = [election.asset['data']]
                 curr_validator_set = bigchain.get_validators(new_height)
                 updated_validator_set = new_validator_set(curr_validator_set,
-                                                          new_height, validator_updates)
+                                                          validator_updates)
 
                 updated_validator_set = [v for v in updated_validator_set if v['voting_power'] > 0]
                 bigchain.store_validator_set(new_height+1, updated_validator_set)
                 return [encode_validator(election.asset['data'])]
         return []
+
+    def _vote_ratio(self, bigchain, height):
+        cast_votes = self._get_vote_ids(bigchain)
+        votes = [(tx['outputs'][0]['amount'], bigchain.get_block_containing_tx(tx['id'])[0]) for tx in cast_votes]
+        votes_cast = [int(vote[0]) for vote in votes if vote[1] <= height]
+        total_votes_cast = sum(votes_cast)
+        total_votes = sum([voter.amount for voter in self.outputs])
+        vote_ratio = total_votes_cast/total_votes
+        return vote_ratio
+
+    def _get_vote_ids(self, bigchain):
+        election_key = self.to_public_key(self.id)
+        votes = get_asset_tokens_for_public_key(bigchain.connection, self.id, election_key)
+        return votes
+
+    def initial_height(self, bigchain):
+        heights = bigchain.get_block_containing_tx(self.id)
+        initial_height = 0
+        if len(heights) != 0:
+            initial_height = min(bigchain.get_block_containing_tx(self.id))
+        return initial_height
+
+    def get_status(self, bigchain, height=None):
```

Comment:
An alternative of recording the election ID and election status with every validator update seems significantly simpler to me. We can extend the `validators` collection or introduce a new `validator_elections` collection to store election statuses.

Recomputing the status is quite complicated and partly repeats what's calculated upon election conclusion.

Reply:
@ldmberman recall that this code is executed only via command line. I think the node operator won't notice a delay of few milliseconds.

Reply:
I do not think performance is a concern either. It's about maintainability.